internal/postgres/report.go

```diff
@@ -0,0 +1,489 @@
+		if syntheticPK != nil {
+			// Warnings about synthetic primary keys must be handled as a special case
+			// because we have a Spanner column with no matching PostgreSQL col.
+			// Much of the generic code for processing issues assumes we have both.
+			if p.severity == warning {
+				l = append(l, fmt.Sprintf("Column '%s' was added because this table didn't have a primary key. Spanner requires a primary key for every table", *syntheticPK))
+			}
+		}
+		issueBatcher := make(map[schemaIssue]bool)
+		for _, pgCol := range cols {
+			for _, i := range issues[pgCol] {
+				if issueDB[i].severity != p.severity {
+					continue
+				}
+				if issueDB[i].batch {
+					if issueBatcher[i] {
+						// Have already reported a previous instance of this
+						// (batched) issue, so skip this one.
+						continue
+					}
+					issueBatcher[i] = true
+				}
+				spCol, err := GetSpannerCol(conv, pgTable, pgCol, true)
+				if err != nil {
+					conv.unexpected(err.Error())
+				}
+				pgType := printType(pgSchema.cols[pgCol])
+				sType := spSchema.Cds[spCol].PrintColumnDefType()
+				// A note on case: Both PostgreSQL types and Spanner types are case
+				// insensitive. PostgreSQL often uses lower case (e.g. pg_dump uses
+				// lower case) and pg_query uses lower case, so pgType is lower-case.
+				// However Spanner defaults to upper case, and the Spanner AST uses
+				// upper case, so sType is upper case. When printing PostgreSQL and
+				// Spanner types for comparison purposes, this can be distracting.
+				// Hence we switch to lower-case for Spanner types here.
+				sType = strings.ToLower(sType)
+				switch i {
+				case defaultValue:
+					l = append(l, fmt.Sprintf("%s e.g. column '%s'", issueDB[i].brief, pgCol))
+				case foreignKey:
+					l = append(l, fmt.Sprintf("Column '%s' uses foreign keys which Spanner does not support", pgCol))
+				case timestamp:
+					// Avoid the confusing "timestamp is mapped to timestamp" message.
+					l = append(l, fmt.Sprintf("Some columns have PostgreSQL type 'timestamp without timezone' which is mapped to Spanner type timestamp e.g. column '%s'. %s", pgCol, issueDB[i].brief))
+				case widened:
+					l = append(l, fmt.Sprintf("%s e.g. for column '%s', PostgreSQL type %s is mapped to Spanner type %s", issueDB[i].brief, pgCol, pgType, sType))
+				default:
+					l = append(l, fmt.Sprintf("Column '%s': type %s is mapped to %s. %s", pgCol, pgType, sType, issueDB[i].brief))
+				}
+			}
+		}
+		if len(l) == 0 {
+			continue
+		}
+		heading := p.heading
+		if len(l) > 1 {
+			heading = heading + "s"
+		}
+		body = append(body, tableReportBody{heading: heading, lines: l})
+	}
+	return body
+}
+
+func fillRowStats(conv *Conv, spTable string, badWrites map[string]int64, tr *tableReport) {
+	rows := conv.stats.rows[spTable]
+	goodRows := conv.stats.goodRows[spTable]
+	badConvRows := conv.stats.badRows[spTable]
+	badRowWrites := badWrites[spTable]
+	if rows != goodRows+badConvRows || badRowWrites > goodRows {
```

Comment:
Why does `badRowWrites > goodRows` indicate inconsistent row counts?

Reply:
Added a comment, and renamed goodRows to goodConvRows.

badRowWrites are rows we converted (hence counted in goodConvRows), but failed to write to Spanner.